### PR TITLE
feat(bridge): per-server preferSharedInstance opt-in (#391)

### PR DESCRIPTION
**Pull request by atusy**
Implements #391 (follow-up to #382/#390): a per-server **`preferSharedInstance`** opt-in that reuses a **single** downstream language-server process across every workspace root — announcing each new root via `workspace/didChangeWorkspaceFolders` — instead of the default one-process-per-marker-root. Honored only when the downstream server advertises `workspace.workspaceFolders.{supported, changeNotifications}`; otherwise it logs once and falls back to per-root instances.

## What's included

- **Config** (`preferSharedInstance: Option<bool>`, default `false`): serde camelCase, overlay-wins merge mirroring `rootMarkers` so the `languageServers._` wildcard is inheritable and a concrete server can opt out of a blanket `_.preferSharedInstance = true`. `config init` emits `preferSharedInstance = false` so the knob is discoverable.
- **Capability detection**: `supports_workspace_folder_changes` reads the stored downstream `ServerCapabilities` (`supported == true` AND `changeNotifications ∈ {true, id}`).
- **Routing**: an explicit `ConnectionKey::Shared` variant (distinct from the client-root fallback). `resolve_acquire` routes opt-in servers to the shared key, but when an existing shared connection is `Ready` and incapable it logs once and diverts new roots to per-root keys — keeping that connection's own spawn root on it so a single root is never split across two processes.
- **Mutable folder set** (`WorkspaceFolderSet`, shared with the reader that answers `workspace/workspaceFolders` pulls): `add_and_announce` couples the folder add with the `didChangeWorkspaceFolders` enqueue under one lock, committing the folder only if the notification queued, ahead of `didOpen` on the single-writer FIFO. A queue-full announce fails the acquire so it retries rather than open a document for an unannounced root.
- The incapable-shared divert and cold spawns are bounded by the caller's (remaining) budget.

## Tests

- Unit: config parse/merge, `ConnectionKey::Shared` distinctness, capability detection (`OneOf` edge cases), `WorkspaceFolderSet` CAS/announce, and `resolve_acquire` routing (shared / capable-join / incapable-fallback / own-root-stays).
- E2E (`tests/e2e_shared_instance.rs`, new `workspace-folders[-incapable]` mock modes): a shared process learns both roots via `didChangeWorkspaceFolders`; the per-root default and an incapable opt-in both isolate roots.

Default behavior (per-root, #382) is unchanged. Docs updated (`docs/README.md`, pool-coordination ADR).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/lsp/bridge/workspace_folders.rs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+            return true;
+        }
+        if announce() {
+            folders.push(folder);
+            true
+        } else {
+            false
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/workspace_folders.rs]: Done in 08b4bd35: `WorkspaceFolderSet` now uses `std::sync::Mutex` (with the project's `recover_poison` helper, not `unwrap()`) and is fully synchronous — `snapshot`/`contains`/`add_and_announce` and `announce_shared_root` are now plain `fn`, so the `.await` here is removed. Thanks!

File: src/lsp/bridge/workspace_folders.rs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        );
+    }
+
+    #[tokio::test]
+    async fn add_and_announce_seeds_a_none_set() {
+        let set = WorkspaceFolderSet::new(None);
+        assert!(set.add_and_announce(folder("file:///a"), || true).await);
+        assert_eq!(set.snapshot().await, Some(vec![folder("file:///a")]));
+    }
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/workspace_folders.rs:163]: Done in 08b4bd35: `WorkspaceFolderSet` now uses `std::sync::Mutex` (with the project's `recover_poison` helper, not `unwrap()`) and is fully synchronous — `snapshot`/`contains`/`add_and_announce` and `announce_shared_root` are now plain `fn`, so the `.await` here is removed. Thanks!

File: src/lsp/bridge/actor/reader.rs [hunk 3]
```diff
@@ -720,11 +721,13 @@ async fn handle_server_request(
...
-                "{}Answering workspace/workspaceFolders from initialize-time folders",
+                "{}Answering workspace/workspaceFolders from the connection's current folders",
                 lang_prefix
             );
-            Ok(serde_json::to_value(deps.workspace_folders.as_ref())
-                .unwrap_or(serde_json::Value::Null))
+            Ok(
+                serde_json::to_value(deps.workspace_folders.snapshot().await)
+                    .unwrap_or(serde_json::Value::Null),
+            )
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/actor/reader.rs]: Done in 08b4bd35: `WorkspaceFolderSet` now uses `std::sync::Mutex` (with the project's `recover_poison` helper, not `unwrap()`) and is fully synchronous — `snapshot`/`contains`/`add_and_announce` and `announce_shared_root` are now plain `fn`, so the `.await` here is removed. Thanks!

File: src/lsp/bridge/pool.rs [hunk 4]
```diff
@@ -577,56 +578,117 @@ impl LanguageServerPool {
...
+        // opening this document on a server that ignores
+        // didChangeWorkspaceFolders would wedge the 2nd+ root. `resolve_acquire`
+        // now sees the shared connection as Ready+incapable, so it re-resolves
+        // to the per-root key; acquire that connection instead, within the
+        // caller's remaining budget.
+        if handle.key().is_shared() && !handle.supports_workspace_folder_changes() {
+            let serves_this_root = match marker.as_ref() {
+                Some((_root, folder)) => handle.workspace_folders().contains(folder).await,
+                None => false,
+            };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs]: Done in 08b4bd35: `WorkspaceFolderSet` now uses `std::sync::Mutex` (with the project's `recover_poison` helper, not `unwrap()`) and is fully synchronous — `snapshot`/`contains`/`add_and_announce` and `announce_shared_root` are now plain `fn`, so the `.await` here is removed. Thanks!

File: src/lsp/bridge/pool.rs [hunk 5]
```diff
@@ -577,56 +578,117 @@ impl LanguageServerPool {
...
+                    .await;
+            }
+        }
+
+        // Announce the (possibly newly-joined) root before the caller opens any
+        // document. Idempotent if `acquire_resolved_wait_ready`'s ReturnExisting
+        // path already announced; on the initializing-retry path it is the only
+        // announce. Propagates a queue-full failure so the caller retries rather
+        // than open a document for an unannounced root.
+        self.announce_shared_root(&handle, &marker).await?;
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs:685]: Done in 08b4bd35: `WorkspaceFolderSet` now uses `std::sync::Mutex` (with the project's `recover_poison` helper, not `unwrap()`) and is fully synchronous — `snapshot`/`contains`/`add_and_announce` and `announce_shared_root` are now plain `fn`, so the `.await` here is removed. Thanks!

File: src/lsp/bridge/pool.rs [hunk 6]
```diff
@@ -681,19 +746,184 @@ impl LanguageServerPool {
...
+                    && !handle.supports_workspace_folder_changes() =>
+            {
+                handle.log_incapable_fallback_once(server_name);
+                // Keep the shared connection's own spawn root on the shared key
+                // (it is already rooted and serving it); divert every other root
+                // to its own per-root process for true per-root isolation.
+                let serves_this_root = match marker.as_ref() {
+                    Some((_root, folder)) => handle.workspace_folders().contains(folder).await,
+                    None => false,
+                };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs]: Done in 08b4bd35: `WorkspaceFolderSet` now uses `std::sync::Mutex` (with the project's `recover_poison` helper, not `unwrap()`) and is fully synchronous — `snapshot`/`contains`/`add_and_announce` and `announce_shared_root` are now plain `fn`, so the `.await` here is removed. Thanks!

File: src/lsp/bridge/pool.rs [hunk 7]
```diff
@@ -681,19 +746,184 @@ impl LanguageServerPool {
...
+        } else {
+            Err(io::Error::new(
+                io::ErrorKind::WouldBlock,
+                format!(
+                    "bridge: could not queue didChangeWorkspaceFolders for {}",
+                    folder.uri.as_str()
+                ),
+            ))
+        }
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs:1001]: Done in 08b4bd35: `WorkspaceFolderSet` now uses `std::sync::Mutex` (with the project's `recover_poison` helper, not `unwrap()`) and is fully synchronous — `snapshot`/`contains`/`add_and_announce` and `announce_shared_root` are now plain `fn`, so the `.await` here is removed. Thanks!

File: src/lsp/bridge/pool.rs [hunk 8]
```diff
@@ -749,9 +980,23 @@ impl LanguageServerPool {
...
+                    None => {
+                        return Err(io::Error::other(
+                            "bridge: connection disappeared before ReturnExisting",
+                        ));
+                    }
+                };
+                // Release the pool lock before the (folder-set lock + send)
+                // announce so it never nests under `connections`.
+                drop(connections);
+                self.announce_shared_root(&handle, &marker).await?;
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs:1071]: Done in 08b4bd35: `WorkspaceFolderSet` now uses `std::sync::Mutex` (with the project's `recover_poison` helper, not `unwrap()`) and is fully synchronous — `snapshot`/`contains`/`add_and_announce` and `announce_shared_root` are now plain `fn`, so the `.await` here is removed. Thanks!

File: src/lsp/bridge/pool.rs [hunk 9]
```diff
@@ -1395,6 +1648,146 @@ mod tests {
...
+            create_handle_with_key(ConnectionState::Ready, ConnectionKey::shared("lua")).await;
+        handle.set_server_capabilities(tower_lsp_server::ls_types::ServerCapabilities::default());
+        // Seed the shared connection's folder set with this doc's marker root,
+        // as if it had spawned rooted there.
+        let (marker, _) = pool.resolve_marker_and_key("lua", &config, Some(&doc));
+        let own_root = marker.expect("doc has a marker root").1;
+        handle
+            .workspace_folders()
+            .add_and_announce(own_root, || true)
+            .await;
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs]: Done in 08b4bd35: `WorkspaceFolderSet` now uses `std::sync::Mutex` (with the project's `recover_poison` helper, not `unwrap()`) and is fully synchronous — `snapshot`/`contains`/`add_and_announce` and `announce_shared_root` are now plain `fn`, so the `.await` here is removed. Thanks!

File: src/lsp/bridge/workspace_folders.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/workspace_folders.rs]: Done in 08b4bd35: `WorkspaceFolderSet` now uses `std::sync::Mutex` (with the project's `recover_poison` helper, not `unwrap()`) and is fully synchronous — `snapshot`/`contains`/`add_and_announce` and `announce_shared_root` are now plain `fn`, so the `.await` here is removed. Thanks!

File: src/lsp/bridge/actor/reader.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/actor/reader.rs]: Done in 08b4bd35: `WorkspaceFolderSet` now uses `std::sync::Mutex` (with the project's `recover_poison` helper, not `unwrap()`) and is fully synchronous — `snapshot`/`contains`/`add_and_announce` and `announce_shared_root` are now plain `fn`, so the `.await` here is removed. Thanks!

File: src/lsp/bridge/pool.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs]: Done in 08b4bd35: `WorkspaceFolderSet` now uses `std::sync::Mutex` (with the project's `recover_poison` helper, not `unwrap()`) and is fully synchronous — `snapshot`/`contains`/`add_and_announce` and `announce_shared_root` are now plain `fn`, so the `.await` here is removed. Thanks!

File: src/lsp/bridge/pool.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs]: Done in 08b4bd35: `WorkspaceFolderSet` now uses `std::sync::Mutex` (with the project's `recover_poison` helper, not `unwrap()`) and is fully synchronous — `snapshot`/`contains`/`add_and_announce` and `announce_shared_root` are now plain `fn`, so the `.await` here is removed. Thanks!

File: src/lsp/bridge/pool.rs [hunk 10]
```diff
@@ -681,19 +746,184 @@ impl LanguageServerPool {
...
+                        target: "kakehashi::bridge",
+                        "[{}] didChangeWorkspaceFolders for {} not queued ({:?}); will retry",
+                        handle.key(),
+                        folder.uri.as_str(),
+                        result
+                    );
+                    false
+                }
+            })
+            .await;
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs]: Done in 08b4bd35: `WorkspaceFolderSet` now uses `std::sync::Mutex` (with the project's `recover_poison` helper, not `unwrap()`) and is fully synchronous — `snapshot`/`contains`/`add_and_announce` and `announce_shared_root` are now plain `fn`, so the `.await` here is removed. Thanks!

File: src/lsp/bridge/pool.rs [hunk 11]
```diff
@@ -749,9 +980,23 @@ impl LanguageServerPool {
...
+                // surface it as an error rather than panicking in production
+                // (no-panic convention) if that ever fails to hold.
+                let handle = match connections.get(&connection_key) {
+                    Some(handle) => Arc::clone(handle),
+                    None => {
+                        return Err(io::Error::other(
+                            "bridge: connection disappeared before ReturnExisting",
+                        ));
+                    }
+                };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs]: Done in 08b4bd35: the `connections` map is now looked up once (`let existing = connections.get(&connection_key)`) and the `ReturnExisting` arm reuses it. Verified it compiles — the `existing` borrow ends before the `SpawnNew` arm's `connections.remove()` under NLL.

File: src/lsp/bridge/workspace_folders.rs [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+        if folders.iter().any(|existing| existing.uri == folder.uri) {
+            return true;
+        }
+        if announce() {
+            folders.push(folder);
+            true
+        } else {
+            false
+        }
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/workspace_folders.rs]: Good catch — fixed in 08b4bd35. `add_and_announce` no longer calls `get_or_insert_with(Vec::new)` before the announce succeeds: a `None` set is materialized into `Some` only on a committed add, so a failed announce leaves the "answer null" state untouched. Added a `failed_announce_keeps_a_none_set_null` regression test.

File: src/lsp/bridge/pool.rs [hunk 13]
```diff
@@ -681,19 +746,184 @@ impl LanguageServerPool {
...
+        document_uri: Option<&Url>,
+    ) -> (
+        Option<(Url, tower_lsp_server::ls_types::WorkspaceFolder)>,
+        ConnectionKey,
+    ) {
+        let (marker, per_root_key) =
+            self.resolve_marker_and_key(server_name, server_config, document_uri);
+        if !server_config.prefers_shared_instance() {
+            return (marker, per_root_key);
+        }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs:838]: Fixed in 08b4bd35. `resolve_acquire` now keeps marker-less documents (no marker root / no hint / non-file URI / `[]` kill switch) on the client-root fallback key instead of the shared key — they have no root to share or announce. The shared key is reserved for marker-rooted docs, matching the `ConnectionRoot::Shared` doc comment. Added `resolve_acquire_keeps_marker_less_docs_on_client_fallback`.

File: src/lsp/bridge/pool.rs [hunk 14]
```diff
@@ -577,56 +578,117 @@ impl LanguageServerPool {
...
+                    .await;
+                return self
+                    .acquire_resolved_wait_ready(
+                        server_name,
+                        server_config,
+                        per_root_key,
+                        per_root_marker,
+                        remaining,
+                    )
+                    .await;
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs]: Done in ca685edc: the divert now builds the per-root key directly from the already-resolved `marker` (no second `resolve_acquire` → no redundant filesystem marker walk or `connections` re-lock), and logs the per-root fallback explicitly so the observability is preserved.

File: tests/test_cli.rs [hunk 15]
```diff
@@ -267,6 +267,30 @@ fn test_config_init_includes_capture_mappings() {
...
+/// `preferSharedInstance = false` under the `languageServers._` wildcard, so
+/// the opt-in (#391) is discoverable in the generated template.
+#[test]
+fn test_config_init_includes_prefer_shared_instance() {
+    let output = Command::new(env!("CARGO_BIN_EXE_kakehashi"))
+        .args(["config", "init"])
+        .output()
+        .expect("Failed to execute command");
+
+    let stdout = String::from_utf8_lossy(&output.stdout);
```
[the remark this answers is not in the document]
Comment (atusy) [reply tests/test_cli.rs:286]: Done in e2214579: the test now asserts `output.status.success()` before checking stdout.

File: src/lsp/bridge/pool.rs [hunk 16]
```diff
@@ -577,56 +578,117 @@ impl LanguageServerPool {
...
+            }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs:677]: Addressed in ca685edc (now e2214579): the divert no longer re-runs `resolve_acquire` — it builds the per-root key directly from the already-resolved `marker`, so there's no filesystem walk / lock between computing `remaining` and the per-root acquire. `remaining` is now accurate.

File: tests/bin/mock_formatter.rs [hunk 17]
```diff
@@ -48,6 +48,13 @@
 //!   `;` as triggers; answers `textDocument/onTypeFormatting` with the
 //!   uppercasing whole-document edit for ANY typed character (bridge-side
 //!   trigger filtering is what `tests/e2e_on_type_formatting.rs` proves).
+//! - `workspace-folders` — advertises `workspace.workspaceFolders.{supported,
+//!   changeNotifications}` + `hoverProvider`; records the `initialize`-time
+//!   workspace folders and every `workspace/didChangeWorkspaceFolders`
+//!   addition, then answers `textDocument/hover` (for opened docs) with the
+//!   sorted list of folder URIs it currently knows. Used by
```
[the remark this answers is not in the document]
Comment (atusy) [reply tests/bin/mock_formatter.rs]: Fixed in e2214579: the mock header now states the `workspace-folders` hover is NOT gated on `didOpen` (the folder set is the subject under test, populated by initialize + didChangeWorkspaceFolders).

File: src/lsp/bridge/pool.rs [hunk 18]
```diff
@@ -577,56 +578,125 @@ impl LanguageServerPool {
...
+        // opening this document on a server that ignores
+        // didChangeWorkspaceFolders would wedge the 2nd+ root. `resolve_acquire`
+        // now sees the shared connection as Ready+incapable, so it re-resolves
+        // to the per-root key; acquire that connection instead, within the
+        // caller's remaining budget.
+        if handle.key().is_shared() && !handle.supports_workspace_folder_changes() {
+            let serves_this_root = match marker.as_ref() {
+                Some((_root, folder)) => handle.workspace_folders().contains(folder),
+                None => false,
+            };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs]: Done in e2214579: simplified with `marker.as_ref().is_some_and(|(_, folder)| handle.workspace_folders().contains(folder))`.

File: src/lsp/bridge/pool.rs [hunk 19]
```diff
@@ -681,19 +754,192 @@ impl LanguageServerPool {
...
+                    && !handle.supports_workspace_folder_changes() =>
+            {
+                handle.log_incapable_fallback_once(server_name);
+                // Keep the shared connection's own spawn root on the shared key
+                // (it is already rooted and serving it); divert every other root
+                // to its own per-root process for true per-root isolation.
+                let serves_this_root = match marker.as_ref() {
+                    Some((_root, folder)) => handle.workspace_folders().contains(folder),
+                    None => false,
+                };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs]: Done in e2214579: simplified with `is_some_and` (the divert site too).

File: src/lsp/bridge/pool.rs [hunk 20]
```diff
@@ -744,14 +991,30 @@ impl LanguageServerPool {
...
+                // surface it as an error rather than panicking in production
+                // (no-panic convention) if that ever fails to hold.
+                let handle = match existing {
+                    Some(handle) => Arc::clone(handle),
+                    None => {
+                        return Err(io::Error::other(
+                            "bridge: connection disappeared before ReturnExisting",
+                        ));
+                    }
+                };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs]: Done in e2214579: `let handle = existing.cloned().ok_or_else(|| io::Error::other(...))?;`.

File: src/lsp/bridge/pool.rs [hunk 21]
```diff
@@ -681,19 +754,192 @@ impl LanguageServerPool {
...
+            });
+
+        if announced {
+            Ok(())
+        } else {
+            Err(io::Error::new(
+                io::ErrorKind::WouldBlock,
+                format!(
+                    "bridge: could not queue didChangeWorkspaceFolders for {}",
+                    folder.uri.as_str()
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs]: Good catch — fixed in e2214579. `announce_shared_root` now maps the send failure to a faithful `io::ErrorKind`: `QueueFull`→`WouldBlock` (retryable), `ChannelClosed`→`BrokenPipe`, `SerializationFailed`→`InvalidData`, mirroring the didClose mapping — so a dead/serialization-failed connection is not retried as transient backpressure.

File: src/lsp/bridge/pool.rs [hunk 22]
```diff
@@ -681,19 +753,208 @@ impl LanguageServerPool {
...
+    ) -> io::Result<()> {
+        if !handle.key().is_shared() {
+            return Ok(());
+        }
+        let Some((_root, folder)) = marker else {
+            return Ok(());
+        };
+        if !handle.supports_workspace_folder_changes() {
+            return Ok(());
+        }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs]: Good catch — fixed in 1fea1006. `announce_shared_root` is now async and holds the `connections` lock across an `Arc::ptr_eq` liveness check AND the add+announce, so a concurrent respawn (get_or_create's SpawnNew purge) can't interleave between validating the live handle and queuing the notification. This mirrors the existing live-handle guard in execute.rs / did_open.rs; a replaced handle returns `BrokenPipe` so the caller re-acquires. `add_and_announce` is synchronous (non-blocking try_send), so no await is held under either lock (order: connections → folder set).

File: src/lsp/bridge/pool.rs [hunk 23]
```diff
@@ -577,56 +578,124 @@ impl LanguageServerPool {
...
+                    .await;
+            }
+        }
+
+        // Announce the (possibly newly-joined) root before the caller opens any
+        // document. Idempotent if `acquire_resolved_wait_ready`'s ReturnExisting
+        // path already announced; on the initializing-retry path it is the only
+        // announce. Propagates a queue-full failure so the caller retries rather
+        // than open a document for an unannounced root.
+        self.announce_shared_root(&handle, &marker)?;
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs]: Done in 1fea1006: this call site now `.await`s `announce_shared_root` (it is async for the under-lock liveness check).

File: src/lsp/bridge/pool.rs [hunk 24]
```diff
@@ -744,14 +1006,25 @@ impl LanguageServerPool {
...
+                // connection exists for this key, so the lookup is an invariant;
+                // surface it as an error rather than panicking in production
+                // (no-panic convention) if that ever fails to hold.
+                let handle = existing.cloned().ok_or_else(|| {
+                    io::Error::other("bridge: connection disappeared before ReturnExisting")
+                })?;
+                // Release the pool lock before the (folder-set lock + send)
+                // announce so it never nests under `connections`.
+                drop(connections);
+                self.announce_shared_root(&handle, &marker)?;
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs]: Done in 1fea1006: `.await`ed here too (the `connections` lock is dropped first, since the tokio mutex is not reentrant and `announce_shared_root` re-locks it).

File: src/lsp/bridge/pool.rs [hunk 25]
```diff
@@ -681,19 +753,208 @@ impl LanguageServerPool {
...
+        let folder = folder.clone();
+        // The closure runs under the folder-set lock and reports its send
+        // outcome out here so the error kind can distinguish retryable
+        // backpressure (queue full) from a dead/serialization-failed connection.
+        let mut send_outcome = NotificationSendResult::Queued;
+        let announced = handle
+            .workspace_folders()
+            .add_and_announce(folder.clone(), || {
+                let result = handle.send_notification(
+                    build_did_change_workspace_folders_notification(vec![folder.clone()]),
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/pool.rs:954]: Not applied: `vec![folder]` would move the closure's captured `folder`, but the following `log::debug!/warn!` lines still use `folder.uri.as_str()`, so `vec![folder]` fails to compile (use-after-move). The `folder.clone()` is required because the folder is logged after the notification is built. Leaving the clone (a cheap `WorkspaceFolder` = one `Uri` + short name).

File: src/lsp/bridge/workspace_folders.rs [hunk 26]
```diff
@@ -0,0 +1,160 @@
...
+    /// observe the folder as present — and so skip its own announce — until the
+    /// first caller's notification is actually on the single-writer FIFO. If the
+    /// announce fails to queue, nothing is committed (and a `None` set stays
+    /// `None`), so a later acquisition re-attempts it rather than opening a
+    /// document for an unannounced root.
+    pub(crate) fn add_and_announce<F>(&self, folder: WorkspaceFolder, announce: F) -> bool
+    where
+        F: FnOnce() -> bool,
+    {
+        let mut guard = self.inner.lock().recover_poison("WorkspaceFolderSet::add");
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/bridge/workspace_folders.rs]: Done in c6490be5: the context string is now `WorkspaceFolderSet::add_and_announce`.

File: src/lsp/lsp_impl/text_document/formatting.rs [hunk 27]
```diff
@@ -907,8 +907,9 @@ async fn dispatch_concatenated_formatting(
                 // The connection `(server, root)` this step's requests route to —
                 // resolved identically to the request path so the scratch-doc
                 // didClose and the per-step cancel target the right process (#382).
-                let connection_key =
-                    pool.resolve_connection_key(&server_name, &server_config, Some(&uri));
+                let connection_key = pool
+                    .resolve_connection_key(&server_name, &server_config, Some(&uri))
+                    .await;
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/text_document/formatting.rs]: Good catch — fixed (folded into the feat(bridge) commit). The concatenated formatting pipeline no longer pre-resolves `connection_key` (which could be the optimistic `shared_key` while a connection is still Initializing). Instead, `pipeline_step_request_kind` now returns the **actually-acquired** `handle.key()` alongside the request kind, and the step uses that for the scratch-doc didClose and (via a new `step_connection_key` OnceLock, mirroring `step_downstream_id`) the timeout cancel. So both target the same connection the request routed to — including a preferSharedInstance server diverted to a per-root connection. Removed the now-unused `resolve_connection_key`. Verified by the formatting unit tests + e2e_concatenated_formatting + e2e_shared_instance.
